### PR TITLE
refactor!: move shop screen to Economy module

### PR DESCRIPTION
**Pull request by ahv15**
Requires https://github.com/Terasology/Economy/pull/22 to test this. Shifts the Shop Screen to the Economy Module so that other gameplay modes can make use of it. Makes use of the walletHUD from the Economy module to display the amount of money a player has instead of moneyLabel in the DefenceHUD.

### CODE REVIEW

File: deltas/engine/prefabs/player/player.prefab [hunk 1]
```diff
@@ -1,5 +0,0 @@
-{
-  "Wallet": {
-    "funds": 100
-  }
```
Comment (skaldarnar) [inline deltas/engine/prefabs/player/player.prefab]: Where did this go, why does the player no longer need funds to start with?

File: module.txt [hunk 2]
```diff
@@ -13,6 +13,10 @@
             "id": "FlexiblePathfinding",
             "minVersion": "1.0.0"
         },
+        {
+            "id": "Economy",
+            "minVersion": "1.1.0-SNAPSHOT"
```
Comment (skaldarnar) [inline module.txt]: We should make a release of the current state of Economy and bump the version number ... :thinking:

File: src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java [hunk 3]
```diff
@@ -15,108 +15,46 @@
...
-        blocks.addAll(blockExplorer.getFreeformBlockFamilies());
-
-        purchasableBlocks = blocks.stream()
-                .map(blockManager::getBlockFamily)
-                .map(BlockFamily::getArchetypeBlock)
-                .filter(block -> block.getPrefab().isPresent())
-                .filter(block -> block.getPrefab().get().hasComponent(PurchasableComponent.class))
-                .collect(Collectors.toSet());
+        player.addOrSaveComponent(component);
+        player.send(new WalletUpdatedEvent(component.amount));
```
Comment (skaldarnar) [inline src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java]: Ah, I see - why was this change necessary? (You probably already explained that before... :see_no_evil: )

File: src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java [hunk 4]
```diff
@@ -132,10 +70,8 @@ private void cleanUpInventory() {
     }
 
     private void resetMoney() {
-        WalletComponent component = assetManager.getAsset(DefenceUris.PLAYER, Prefab.class)
-                .map(prefab -> prefab.getComponent(WalletComponent.class))
-                .orElse(new WalletComponent());
-        localPlayer.getCharacterEntity().addOrSaveComponent(component);
+        localPlayer.getCharacterEntity().getComponent(CurrencyStorageComponent.class).amount = 100;
+        localPlayer.getCharacterEntity().send(new WalletUpdatedEvent(100));
```
Comment (skaldarnar) [inline src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java]: I've seen the magic number `100` a couple of times here now. I think it makes sense to turn this value into a constant to be used in all the places.

**Review by skaldarnar — COMMENTED** [review]
I didn't test this in-game yet.

@syntaxi in case you happen to have some time, maybe you can have a look at these changes trying to make the shop+wallet functionality available to other modules as well.

File: overrides/Inventory/ui/ingame/inventoryScreen.ui [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "type": "shopScreen",
-  "skin": "DefenceDefault",
```
Comment (jdrueckert) [inline overrides/Inventory/ui/ingame/inventoryScreen.ui:3]: I don't see the actual skin file being removed?

**Review by jdrueckert — CHANGES_REQUESTED** [review]
Would it make sense to move the actual shop screen ui file to Economy as well? I guess the structure is the same between GD and LaS and only the items available in the shop would differ, right?

File: src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java [hunk 6]
```diff
@@ -15,108 +15,47 @@
...
  * Handles the purchasing of blocks
  */
 @RegisterSystem
-@Share(ShopManager.class)
 public class ShopManager extends BaseComponentSystem {
-
-    private Set<Block> purchasableBlocks = new HashSet<>();
-    private Set<Prefab> purchasableItems = new HashSet<>();
-
+    int baseAmount = 100;
```
Comment (syntaxi) [inline src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java]: I would prefer to leave this as a prefab, as it would allow the value to be changed

File: src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java [hunk 7]
```diff
@@ -15,108 +15,47 @@
...
-        }
-    }
-
-    @Override
-    public void postBegin() {
-        blockItemFactory = new BlockItemFactory(entityManager);
-        BlockExplorer blockExplorer = new BlockExplorer(assetManager);
+    @ReceiveEvent(priority = EventPriority.PRIORITY_LOW)
+    public void onPlayerJoin(OnPlayerSpawnedEvent onPlayerSpawnedEvent, EntityRef player) {
+        CurrencyStorageComponent component = new CurrencyStorageComponent();
```
Comment (syntaxi) [inline src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java]: Like my prior comment, can't this component be added to the player in the prefabs.
This change would also allow for other modules to add money to entities, without having to do so through the code every time

File: src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java [hunk 8]
```diff
@@ -132,10 +71,8 @@ private void cleanUpInventory() {
     }
 
     private void resetMoney() {
-        WalletComponent component = assetManager.getAsset(DefenceUris.PLAYER, Prefab.class)
-                .map(prefab -> prefab.getComponent(WalletComponent.class))
-                .orElse(new WalletComponent());
-        localPlayer.getCharacterEntity().addOrSaveComponent(component);
+        localPlayer.getCharacterEntity().getComponent(CurrencyStorageComponent.class).amount = baseAmount;
```
Comment (syntaxi) [inline src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java]: The method already in the resetMoney function can be tweaked to allow for base values to be set in the prefab.
Alternatively, this "reset" function could be moved to the Economy module, however this would depend on your design, which I am not familiar with

**Review by syntaxi — COMMENTED** [review]
Did a brief look over.
Because it's moving components as opposed to replacing it I didn't have too many things to say

File: overrides/Inventory/ui/ingame/inventoryScreen.ui (hunk 5 above)
Comment (ahv15) [reply overrides/Inventory/ui/ingame/inventoryScreen.ui:3]: It's being used in some other UI files and is not specific to this file.

File: src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java (hunk 7 above)
Comment (ahv15) [reply src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java]: The only reason I did this was because Economy creates the CurrencyStorageComponent and adds it to the player by default overriding the defined CurrencyStorageComponent in the player's prefab. Doing it in this manner allows us to make use of a value which will always be constant rather than always changing.

Comment (ahv15) [reply src/main/java/org/terasology/gooeyDefence/economy/ShopManager.java]: Just saw the comment below I got what you meant.

File: deltas/engine/prefabs/player/player.prefab [hunk 9]
```diff
@@ -0,0 +1,5 @@
+{
+  "CurrencyStorage": {
+    "amount": 100
+  }
+}
```
Comment (jdrueckert) [inline deltas/engine/prefabs/player/player.prefab]: Why do we need this in GD, but not in LaS?

Comment (ahv15) [reply deltas/engine/prefabs/player/player.prefab]: Because LaS makes use of the default amount in Economy but GooeyDefence needs to override it.

Comment (jdrueckert) [reply deltas/engine/prefabs/player/player.prefab]: That means that as soon as somebody changes the default in Economy (a module that itself should be unaware of the modules using it), this will break the balancing of LaS. That sounds like a bad idea. Please also add the override to LaS, even if it's set to the same value as the (current) default in Economy.

Comment (ahv15) [reply deltas/engine/prefabs/player/player.prefab]: I made the changes but I couldn't test this in multiplayer as it is taking over 30 minutes for multiplayer to load.

Comment (ahv15) [reply deltas/engine/prefabs/player/player.prefab]: It finally loaded but for some reason there seems to be only a single CurrencyStorageComponent shared among all the players. So any item purchased by one player causes a reduction in other players wallet too. I'll try and fix this but it is taking a lot of time for multiplayer to load currently.

**Review by jdrueckert — APPROVED** [review]
